### PR TITLE
move designer attribute back to document level scanner but made it to…

### PR DESCRIPTION
**Pull request by heejaechang**
… use keep alive connection rather than creating new connection every time. also kept some optimization such as checking cps projects or not reporting as much as we can.

### CODE REVIEW

**Comment by heejaechang** [comment]
@dotnet/roslyn-analysis @dotnet/roslyn-ide can you take a look?

**Comment by heejaechang** [comment]
ping?

**Comment by CyrusNajmabadi** [comment]
> move designer attribute back to document level scanner 

What's the reason for this?  The reason we moved it to being project based was that document based was so incredibly chatty.  It was better to be able to process each project in one large chunk each.

File: src/VisualStudio/Core/Def/Implementation/DesignerAttribute/DesignerAttributeState.cs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+                    // nothing is actually updated
+                    return false;
+                }
+
+                // value updated
+                _lastReported[id] = designerAttributeArgument;
+                return true;
+            }
+
+            protected override string StateName
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/DesignerAttribute/DesignerAttributeState.cs]: nit: use =>

**Comment by heejaechang** [comment]
it was too slow. some people complaining they can't open winform file in designer since cs is not marked as a form for a long time.

**Comment by CyrusNajmabadi** [comment]
Hrmm... that should only be a problem the first time around.  Once hte data is computed and cached then things will be ok in the future (As this data is persisted across VS sessions).

I am worried about making it so chatty again as that just ended up causing massive UI issues.  It seems like we'll just be trading one problem here (slow icon appearing) for another (huge UI delays).  The latter issue seems much worse to me.

**Comment by heejaechang** [comment]
when you say chatty, are you talking about talking to OOP? or jumping to UI thread?

for OOP part, code lens and todo comments, highlight reference is as chatty as this one.

for UI thread, it is using foreground notification service which will batch up multiple requests to 1 UI call, so I dont believe it will cause any UI issues.

**Comment by heejaechang** [comment]
ping @mavasani can you take a look?

**Comment by sharwell** [comment]
@heejaechang At a high level, can you describe why a document-based scanner should produce results in less time than the current approach (solution- or project-based?) ?

**Comment by heejaechang** [comment]
solution cralwer runs all document based one first and then project one. project one is basically one solution crawler run if it sures that it is okay to run very very late.

**Comment by sharwell** [comment]
@heejaechang Awesome, that makes perfect sense. Hopefully we can find a balance between the two; let me review this and get back to you (feel free to nag me about it if you don't hear back).

**Comment by CyrusNajmabadi** [comment]
Sorry, this part wasn't addressed:

> Hrmm... that should only be a problem the first time around. Once hte data is computed and cached then things will be ok in the future (As this data is persisted across VS sessions).

It's unclear to me why this isn't sufficient.  The only time this should be slow is when truly creating a form for the first time.  Normally once computed, the data will be cached, and doing the project form will be sufficient. 

I'm still very hesitant about this moving to be a document scanner.  The UI hit was *really bad*.  Like stupendously bad.  Like "i can't use roslyn for a minute because the UI thread is being hammered" bad.  

If people are complaining that it's taking too long after making a new form for it to show up as a form, then i would prefer we just attack that problem.  i.e. enqueue a document analysis request only when a new document appears.

**Comment by CyrusNajmabadi** [comment]
Again, my primary concern here is that cure may be worse than the disease.  I absolutely think addressing slow form detection should be something done.  But a solution here should really be crafted to not regress things.

--

Note: i would also suggest that rather than just rolling back here, we go the route of a true 'background-async-only' approach for this entire problem space.  i.e. remove the UI thread dependency entirely here.  Then, this feature can safely move to being a document-analyzer and we can be very chatty.  This would require work from teh project system side.  But that's something that we've known needs to be done anyways as this UI thread bottle-neck design has always been a major problem.  It would be better to just invest in fixign the space properly IMO than doing a half measure that reverts us back to where we were.

**Comment by heejaechang** [comment]
@CyrusNajmabadi I am not sure where you got UI hit was so bad data. designer scanner was like that since beginning of Roslyn. we had allocation and etc perf issues since it created compilations for whole solution which should be addressed by moving to OOP. UI issues started to happen once we moved to some projects to cps, and designer scanner is completely turned off for cps projects. so it shouldn't cause the issue in cps anymore. also, cps causing UI issues for designer attribute was cps's bug not designer attribute.

foreground notification should batch UI calls as well, so I dont believe there is jumping issue. tagger does jumping way more than designer scanner. and we are fine with tagger doing the UI switching.

about why it is matter being slow, I dont know, user might removed suo file and open the solution?

**Comment by heejaechang** [comment]
anyway, we have RPS test to catch issues if it is really an issue. and we will have data then to precisely what to address if there is issue.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi I am not sure where you got UI hit was so bad data. designer scanner was like that since beginning of Roslyn.

Because it was actually affecting the team.  We profiled and the traces showed huge UI delays in roslyn due to this.  

> designer scanner was like that since beginning of Roslyn.

Yes.  And it was bad.  And as roslyn grew in size it got worse and worse.  The change to make this project-based was very explicitly done because of the problem here.  

--

So, to repeat: there are two problems that are important enough to need a good fix:

1. Being document based causes massive UI thread contention, making for an extremely bad experience with large projects.
2. Slow analysis causes forms to not show up too quickly.

We need a solution for both, not just for one.  Switching to making '2' good at the expense of '1' just means we still have a bug.

> anyway, we have RPS test to catch issues if it is really an issue.

RPS was not catching this issue.   But it still existed.  We can't just potentially regress perf a lot and then ignore the issue just because our existing perf infrastructure doesn't catch it.  We need a good solution for the whole problem.

**Comment by heejaechang** [comment]
"Being document based causes massive UI thread contention, making for an extremely bad experience with large projects."

this is the part I dont see. are you there you are not refering to cps? if it was, we should have UI delay watson report for a long time. but we didn't until the cps project move.

**Comment by CyrusNajmabadi** [comment]
> foreground notification should batch UI calls as well, 

Perhaps this just wasn't happening.  And perhaps that's a suitable approach here.  Just make things a document scanner, but actually *ensure* that enough batching is happening.  At the time, if batching was happening it was not enough as the UI thread was completely saturated with update calls, and that was showing up in ETW as actually making the UI thread unresponsive for large periods of time.

--

Note: i'm worried that we'll think things are ok due to the analysis not happening for CPS projects.  NOw htat we've moved majorly over to most (all) of the projects being CPS, it sounds like we might flip this switch and then think things are ok since roslyn won't actually do anything for these projects.

The right way to validate this would be to switch roslyn to analyzing CPS projects, and then just swap out the call to update the project with a no-op call that blocks the UI thread for some amount of time.  Then see if this can be switched on in roslyn without causing a problem.

**Comment by CyrusNajmabadi** [comment]
> if it was, we should have UI delay watson report for a long time.

Each UI operation was not long enogh to trigger the delay.  The problem was that the UI thread was completely saturated with these.  i.e. it wasn't 1 operation that was taking seconds.  It was that we had thousands of operations taking a few ms.  Combined with all the rest of the work being done on the UI thread it made typing and otherwise trying to use VS impossible.

@sharwell and i both encountered this many months back and this was one of the things done to help ameliorate the situation.  

--

Again, i'm fine with a change like this, as long as it doesn't regress the situation for people with large non-CPS projects.  We can try simulating that sort of experience ourselves with roslyn to make sure we're not rolling back that problem.  All we need to do is remove the blocking code that stops analysis if htis is a CPS project, and instead of calling into the project system to update things, calling into a small stub that blocks the UI for a small period of time.

**Comment by heejaechang** [comment]
but doing the work at project level, doesnt change this.

" it wasn't 1 operation that was taking seconds. It was that we had thousands of operations taking a few ms. Combined with all the rest of the work being done on the UI thread it made typing and otherwise trying to use VS impossible."

once project level analysis is done, it has to do same work on UI thread to notify the changes. since unless there is a bug, 2 should result in same amount of work to do on UI thread.

**Comment by CyrusNajmabadi** [comment]
> once project level analysis is done, it has to do same work on UI thread to notify the changes. since unless there is a bug, 2 should result in same amount of work to do on UI thread.

In the project form, literally all changes to the project system were batched into one UI call with RegisterDesignerAttributes.  Doing one UI call was a lot better for responsiveness than having thousands of little delays.  THe overhead of each really adds up to a huge amount of time that VS is not usable.

**Comment by heejaechang** [comment]
that is redundant and actually bad. whole point of foreground notification service is so that one doesn't need to do that.

http://source.roslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Implementation/ForegroundNotification/ForegroundNotificationService.cs,c22a46be8182cf31

what existing code does can lock up UI thread for long time in worst case. and calling notification service  100 times doesn't mean it switch to UI thread 100 times. everything is done BG.

the service exact pending request in 50ms interval as long as there is no UI/input request pending. and execute pending work in batch. and pause UI work if there is input request pending.

**Comment by CyrusNajmabadi** [comment]
Again, what i care most about is that we do not regress UI performance.  However that is accomplished is fine with me.  The core problem is that the previous model did not provide this behavior.  Large non-CPS solutions incurred pretty terrible UI delays.  All i want is that we don't fix one problem by reintroducing another.

> that is redundant and actually bad. whole point of foreground notification service is so that one doesn't need to do that.

That is not how it was actually working in practice.  Perhaps because each individual request was taking long enough that all successive requests still ended up saturating and making the UI thread unusable.

**Comment by heejaechang** [comment]
for cps case, my understanding was, cps took long time unlike legacy project to respond to design attribute report. and I believe that was a bug in cps, but since that might take long time to fix, and we need to use VS for Roslyn, I believe so we made changes on our side.

but batching it to project won't improve the fact that cps is slow, it probably (1) block for a long time once, and existing behavior blocks it multiple times less than (1) but longer enough to be annoying multiple times.

so we ends up blocking whole cps projects.

but that's cps, we had this code for long time, and I don't remember any report (we have solution load RPS tests, we have UI delay watson report from wild) due to the designer scanner.

we did have allocations regression due to designer scanner or cpu regression due to designer scanner. but not UI delay though.

**Comment by heejaechang** [comment]
ping?

**Comment by heejaechang** [comment]
closed and re-open to run tests.

**Comment by heejaechang** [comment]
retest please

**Comment by heejaechang** [comment]
retest test please

**Comment by heejaechang** [comment]
retest vsi please

**Comment by heejaechang** [comment]
@dotnet/roslyn-infrastructure how can I retest vsi tests?

**Comment by heejaechang** [comment]
@sharwell have you had chance to take a look? @dotnet/roslyn-analysis @dotnet/roslyn-ide ping?

**Comment by sharwell** [comment]
I talked with @heejaechang today about the primary two issues mentioned by @CyrusNajmabadi:

1. **UI thread issues:** This operation uses a service which batches work on the UI thread, so multiple calls to perform UI work do not necessarily result in multiple items getting scheduled in the dispatcher. The conclusion made by HeeJae (and for which I see no counter evidence) is the original UI thread problems were resolved by #19270 _after_ #19194 was already merged. Had these been implemented the other way around, we likely would have realized that #19194 was not necessary to resolve our prominent UI thread delays. Furthermore, the incrementality of UI thread requests helps prevent long operations from delaying user input.
2. **RPC overhead:** The current belief is the primary performance overhead was the creation of connections. Our new keepalive connection should not suffer from the same problems. While this is an area we are actively working to improve (reduce overhead, increase throughput, reduce latency), there is no reason to believe that the negative impact of multiple RPC calls would outweigh the advantages we get from processing these files at a document granularity.

File: src/VisualStudio/Core/Def/Implementation/DesignerAttribute/DesignerAttributeIncrementalAnalyzer.cs [hunk 2]
```diff
@@ -53,198 +50,190 @@ internal partial class DesignerAttributeIncrementalAnalyzer : ForegroundThreadAf
...
+            var workspace = document.Project.Solution.Workspace as VisualStudioWorkspaceImpl;
+            if (workspace == null)
             {
-                // it is no longer part of the solution
                 return;
             }
 
-            if (ErrorHandler.Succeeded(vsDocument.Project.Hierarchy.GetProperty(itemId, (int)__VSHPROPID.VSHPROPID_ItemSubType, out var currentValue)))
+            var documentId = document.Id;
+            _notificationService.RegisterNotification(() =>
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Implementation/DesignerAttribute/DesignerAttributeIncrementalAnalyzer.cs]: :bulb: We should consider submitting items where `designerAttributeArgument` is not null or empty at higher priority than the rest, to minimize the delay from solution load to when designers can be used.

**Comment by jinujoseph** [comment]
test windows_release_vs-integration_prtest

**Comment by jinujoseph** [comment]
test windows_debug_vs-integration_prtest

**Pull request merged**